src/cargo/core/compiler/job_queue.rs

```diff
@@ -887,19 +886,14 @@ impl<'cfg> DrainState<'cfg> {
     fn handle_error(
         &self,
         shell: &mut Shell,
-        err_state: &mut Option<anyhow::Error>,
+        err_state: &mut ErrorsDuringDrain,
         new_err: anyhow::Error,
     ) {
-        if err_state.is_some() {
-            // Already encountered one error.
-            log::warn!("{:?}", new_err);
-        } else if !self.active.is_empty() {
-            crate::display_error(&new_err, shell);
+        crate::display_error(&new_err, shell);
+        if !self.active.is_empty() && err_state.count == 0 {
             drop(shell.warn("build failed, waiting for other jobs to finish..."));
-            *err_state = Some(anyhow::format_err!("build failed"));
-        } else {
-            *err_state = Some(new_err);
         }
```

Comment:
In response to:

> I believe the original intention of this code was, when possible, to return upwards the error we encountered during compilation if only one was encountered. That being said, however, I don't believe we can reliably return upwards an error so I think it would be better to unconditionally, for any errors encountered during the build, to use `crate::display_error` and print them here. That way there's much less juggling of trying to preserve some "main" error to return while also ensuring that all other errors are printed.